sip/sip-002-smart-contract-language.md

```diff
@@ -0,0 +1,535 @@
+# Abstract
+
+In order to support applications which require validation of some
+pieces of their logic, we present a smart contracting language for use
+with the Stacks blockchain. This smart contracting language can be
+used on the Stacks blockchain to support programatic control over
+digital assets within the Stacks blockchain (e.g., BNS names, Stacks
+tokens, etc.)
+
+While application-chains may use any smart-contract language that they
+like, this smart contracting language's VM will be a part of
+blockstack-core, and, as such, any blockstack-core node will be able to
+validate application chains using this smart contracting language with
+a simple configuration change.
+
+This smart contracting language permits static analysis of any legal
+smart contract to determine runtime costs. This smart contracting
+language is not only Turing-incomplete (a requirement for such static
+analysis to be guaranteed successful), but readily permits other kinds
+of proofs to be made about the code as well.
+
+# Design
+
+A smart contract is composed of two parts:
+
+1. A data-space, which is a set of tables of data which only the
+   smart contract may modify
+2. A set of transactions which operate within the data-space of the
+   smart contract, though they may call transactions from other smart
+   contracts.
```

Comment:
Was hoping to get some clarity on this.  Can contracts send transactions "internally"?  Why is this the right abstraction?  Specific to this draft, why do we need contracts to send first-class transactions when we have `as-contract` and `contract-call`?

Reply:
I think my confusion comes from trying to map the concepts of "client sends a request to the blockchain" and "contract A invokes contract B" onto "blockstack-core keeps an append-only log of all writes" when it comes to a hypothetical implementation.  Let me see if I understand the lifecycle of a client-submitted transaction correctly:

* let tx = the client's request
* if tx does not create a contract or invoke a contract (i.e. it's a Stacks-token-only operation), then handle it accordingly (check that the tx fee is high enough, check that the sender has enough tokens, etc.).
* if tx invokes a contract, then do the following:
   * get the balance of the sender's account.  If it's less than the tx fee, then RETURN INVALID.
   * debit the user's account the tx fee.
   * look up the contract by hash.  If it does not exist, then RETURN INVALID.
   * look up the contract's `define-tx` method and compare the tx's arguments against it.  If the tx does not call an existing method, or supplies invalid arguments, then RETURN INVALID.
   * proceed to execute the transaction code, assigning a fee for each instruction processed.  If running the next instruction would cause the required fee to exceed the tx's fee before the code terminates, then RETURN INVALID.
   * commit the effects of running the code and RETURN OK
* otherwise -- the tx creates a contract
   * check the user's balance.  If zero, then RETURN INVALID
   * check the tx fee against the user's balance. If it's higher, then RETURN INVALID
   * debit the tx fee from the user's balance.
   * check the syntax, calculating the fee of verifying each code item.  If the cost of checking the next item exceeds the tx fee, or if the syntax is invalid, then RETURN INVALID.
   * build the AST, and assign a fee for adding each AST item.  If the cost of adding the next item to the tree exceeds the tx fee (or if the AST gets too big), then RETURN INVALID.
   * walk the AST to check all contract calls. Each step in the walk incurs a small fee.  Do the following while the tx fee is higher than the total cost incurred by walking to the next node in the AST:
      * if the next node calls a contract method, then verify that the contract exists and the method arguments match the contract's `define-tx` method's signature.  If not, then RETURN INVALID.
      * follow a contract call and add its operations to the AST (assigning a fee for each new operation, as before), thereby building up a "super-AST" that includes all instructions in the transitive closure of the call graph.
          * If a smart contract is visited more than once (i.e. there exists a cycle), then RETURN INVALID.
          * If the "super-AST" gets too big, then RETURN INVALID.
          * If the cost of processing another item in the AST exceeds the tx fee, then RETURN INVALID
   * walk this contract's AST to find all tables that need to exist.  Each step in the walk incurs a small fee.
   * create all the tables if the cost of creating them is smaller than the remaining tx fee.  If not, then RETURN INVALID.

I think this makes it so the VM only has to do heavy analysis when a new contract is submitted.  But I don't understand why the act of invoking a method in another smart contract is itself considered to be the same kind of object with the same kind of effects as a client request?

Reply:
Yes, that cycle makes sense (though we wouldn't build a "super-AST" -- there's no need to inline a cross-contract call, however it could be implemented either way).

> But I don't understand why the act of invoking a method in another smart contract is itself considered to be the same kind of object with the same kind of effects as a client request?

I'm not following this question.